### PR TITLE
add upgrade unit tests & migration script

### PR DESCRIPTION
**Pull request by volodymyr-basiuk**

### CODE REVIEW

File: test/state/stateV2_v3_migration.test.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    },
+    {
+      id: '6901746346790563787434755862277025452451108972170386555162524223864832',
+      oldState: '2199023255552',
+      newState: '3298534883328',
+      isOldStateGenesis: false,
+    },
+  ];
+
+describe.only("Deploy StateV2 v2 and migrate", () => {
```
Comment (AndriianChestnykh) [inline test/state/stateV2_v3_migration.test.ts:29]: do you need "only" here?

File: test/state/stateV2_v3_migration.test.ts [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    await expect(stateV3.transitStateGeneric(
+        stateTransitionsWithNoProofs[0].id, 
+        stateTransitionsWithNoProofs[0].oldState, 
+        stateTransitionsWithNoProofs[0].newState,
+        stateTransitionsWithNoProofs[0].isOldStateGenesis,
+        1,
+        []
+      )).to.be.revertedWith(
+      "Default Id Type is not initialized"
+    );
```
Comment (AndriianChestnykh) [inline test/state/stateV2_v3_migration.test.ts:80]: The tests here should kind of a smoke tests. I don't think there is any reason to check anything except a single positive case.
The rest negative and positive tests should already exist at this smart contract regular unit tests.

File: scripts/upgrade-test-def-id-type.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+*/
+
+async function main() {
+
+    const signers = await ethers.getSigners();
+    const stateDeployHelper = await DeployHelper.initialize(null, true);
+    const stateContractMigrationHelper = new StateContractMigrationHelper(stateDeployHelper, signers[0]);
+
+    const oldContractABI = require("./StateV2_deployed_abi.json");  // abi of contract that will be upgraded
+    const stateContractAddress = "0x316345cC62A81bfD5dAe4f37FFBD07FD95b4f51c";  // address of contract that will be upgraded
```
Comment (AndriianChestnykh) [inline scripts/upgrade-test-def-id-type.ts:20]: Better to leave it empty to make people explicitly insert contracts here at before the run.

File: test/state/stateV2_v3_migration.test.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    const stateContractMigrationHelper = new StateContractMigrationHelper(deployHelper, signers[0]);
+    const oldContractABI = require("../../scripts/StateV2_deployed_abi.json");  // abi of contract that will be upgraded
+    const stateContractAddress = state.address;  // address of contract that will be upgraded
+    const stateContractInstance = await stateContractMigrationHelper.getInitContract({
+        contractNameOrAbi: oldContractABI,
+        address: stateContractAddress,
+    });
+
+    // 2. publish first state
+    const params1 = await publishState(stateContractInstance, stateTransitionsWithProofs[0]);
```
Comment (AndriianChestnykh) [inline test/state/stateV2_v3_migration.test.ts:55]: Why don't we extend this method (or add another method) in state utils to working with generic state transition?

File: test/state/stateV2_v3_migration.test.ts (hunk 1 above)
Comment (volodymyr-basiuk) [reply test/state/stateV2_v3_migration.test.ts]: no, removed

File: test/state/stateV2_v3_migration.test.ts (hunk 4 above)
Comment (volodymyr-basiuk) [reply test/state/stateV2_v3_migration.test.ts:55]: The `publishState` function knows how to work with data input (prepare inputs for transaction) we have on separate json files and we use it in others unit tests. I guess we need to reuse that logic here as well.

File: contracts/test-helpers/IState_deployed.sol [hunk 5]
```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity 0.8.16;
+
+uint256 constant MAX_SMT_DEPTH = 64;
```
Comment (AndriianChestnykh) [inline contracts/test-helpers/IState_deployed.sol:4]: The approach to move the contract here is not good enough.
What about it's dependencies: SMT, etc.
Where is the guarantee it is not changed by this time?
And what about the deploy scripts that were used? I see you've copied them but does not look scalable.
Let's discuss

File: helpers/DeployHelper.ts [hunk 6]
```diff
@@ -61,16 +61,24 @@ export class DeployHelper {
...
-    const stateV2 = await upgrades.deployProxy(StateV2Factory, [verifier.address, NetworkIdTypes.polygonMumbai], {
-      unsafeAllowLinkedLibraries: true,
-    });
+    let stateV2;
+    if (isDeployedStateContract) {
+      stateV2 = await upgrades.deployProxy(StateV2Factory, [verifier.address], {
+        unsafeAllowLinkedLibraries: true,
+      });
+    } else {
+      stateV2 = await upgrades.deployProxy(StateV2Factory, [verifier.address, NetworkIdTypes.polygonMumbai], {
```
Comment (AndriianChestnykh) [inline helpers/DeployHelper.ts:77]: Why don't we use hardcoded NetworkIdTypes.polygonMumbai?
What if I would deploy it to other network? Will it go with other id type?

**Review by AndriianChestnykh — CHANGES_REQUESTED** [review]

**Review by AndriianChestnykh — CHANGES_REQUESTED** [review]

**Review by AndriianChestnykh — CHANGES_REQUESTED** [review]

**Review by AndriianChestnykh — APPROVED** [review]

**Pull request merged**